### DISCUSSION TITLE
Determine if an action is asynchronous

### DISCUSSION THREAD

**Discussion by Slooowpoke** [mobxjs/mobx-state-tree#1855] [Q&A]
I’d like to be able to access a store object and determine if an action is asynchronous or not. I wasn’t sure if this is possible?

**Comment by EmilTholin** [comment]
Hi @Slooowpoke!

Do you have an example where it would be beneficial for you to know if the action is asynchronous or not? There might be another way to approach the problem.

**Reply by Slooowpoke** [reply]
Hi @EmilTholin, thanks for answering. Its a little weird but:

I've been working on using xstate with mobx-state tree (https://github.com/mobxjs/mobx-state-tree/issues/1149). One of the problems I encountered was I have to map actions, services and guards across to xstate config which grew pretty tedious (since I've already defined all the mobx actions, and technically by wrapping them in flow defined them as async).

Here's an example of the mapping I have to do:
```
.volatile(self => ({
    xstate: createXStateStore(machineDefinition, {
      actions: {
        fetchData: self.fetchData,
        updateData: self.updateData,
        showErrorMessage: self.showErrorMessage
      },
      services: {
        someAsyncAction: self.someAsyncAction,
      },
      guards: {
        someMobxView: self.someMobxView
      }
    })
  }))
```

I ended up with an okay solution, which was suffixing all the actions with the type (service for async, action for non async), then inside ``createXStateStore`` parsing these keys.

Actions look like this in Mobx-State-Tree:
```
'service.someAsyncAction': () => flow...
'action.regularAction: () => // normal actions
```

Then I collect all the members from the store:
```
const members = getMembers(store) // getMembers from mobx

const { actions, views } = members;

// Split actions based on their name
let asyncActions = {};
let regularActions = {}
Object.keys(actions).map((key) => // if key contains service, add to async, otherwise add to regularActions

const xstateConfig = {
 services: asyncActions,
 actions: regularActions,
 guards: views
}
```

Again this wasn't an amazing solution, since I have to add 'service' or 'action' to everyone, I added some helpers which do this but the whole thing feels really hacky because of all the string parsing and key checking. If there was a way that I could determine from getMembers(store) whether the action was asynchronous or not I could drop the string suffixing and just pass them directly in (like the views).

Let me know if I can explain more or if I'm going about this the wrong way entirely.
